src/AppBundle/Security/ApiKeyAuthenticator.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace AppBundle\Security;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Authentication\Token\PreAuthenticatedToken;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\Authentication\SimplePreAuthenticatorInterface;
+use Doctrine\Common\Persistence\ManagerRegistry;
+
+class ApiKeyAuthenticator implements SimplePreAuthenticatorInterface
+{
+    /**
+     * @var ManagerRegistry
+     */
+    private $registry;
+
+    /**
+     * @param ManagerRegistry $registry
+     */
+    public function __construct(ManagerRegistry $registry)
+    {
+        $this->registry = $registry;
+    }
+
+    /**
+     * @param Request $request
+     * @param string $providerKey
+     * @return PreAuthenticatedToken
+     */
+    public function createToken(Request $request, $providerKey)
+    {
+        $apiKey = $request->headers->get('API-Key-Token');
+        if (!$apiKey) {
+            throw new BadCredentialsException();
+        }
+
+        return new PreAuthenticatedToken(
+            'customer',
+            $apiKey,
+            $providerKey
+        );
+    }
+
+    /**
+     * @param TokenInterface $token
+     * @param string $providerKey
+     * @return bool
+     */
+    public function supportsToken(TokenInterface $token, $providerKey)
+    {
+        return $token instanceof PreAuthenticatedToken && $token->getProviderKey() === $providerKey;
+    }
+
+    /**
+     * @param TokenInterface $token
+     * @param UserProviderInterface $userProvider
+     * @param string $providerKey
+     * @return PreAuthenticatedToken
+     */
+    public function authenticateToken(TokenInterface $token, UserProviderInterface $userProvider, $providerKey)
+    {
+        if (!$userProvider instanceof ApiKeyUserProvider) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'The user provider must be an instance of ApiKeyUserProvider (%s was given).',
+                    get_class($userProvider)
+                )
+            );
+        }
+
+        $apiKey = $token->getCredentials();
+        $username = $userProvider->getUsernameByApiKey($apiKey);
+
+        if (!$username) {
+            // CAUTION: this message will be returned to the client
+            // (so don't put any un-trusted messages / error strings here)
+            throw new CustomUserMessageAuthenticationException(
```

Comment:
CustomUserMessage? really? What this exception about? Rename to understandable name of exception